File: website/src/__generated/services/MemberProfileController.ts
import type { Executor } from '../';
import type { MemberProfileDto } from '../model/dto';
import type { MemberProfileInput, Page, Unit } from '../model/static';

export class MemberProfileController {
    
    constructor(private executor: Executor) {}
    
    async findComplexProfiles(options: MemberProfileControllerOptions['findComplexProfiles']): Promise<
        Page<MemberProfileDto['DEFAULT']>
    > {
        let _uri = '/members/profiles/';
        let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
        let _value: any = undefined;
        _value = options.memberProfileInput.avatar;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'avatar='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.memberProfileInput.bilibili;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'bilibili='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.memberProfileInput.birthday;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'birthday='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.memberProfileInput.description;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'description='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.memberProfileInput.email;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'email='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.memberProfileInput.github;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'github='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.memberProfileInput.id;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'id='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.memberProfileInput.location;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'location='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.memberProfileInput.memberId;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'memberId='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.memberProfileInput.nickname;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'nickname='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.memberProfileInput.roleId;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'roleId='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.memberProfileInput.website;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'website='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.page;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'page='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.size;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'size='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        return (await this.executor({uri: _uri, method: 'GET'})) as Page<MemberProfileDto['DEFAULT']>
    }
    
    async findFullProfile(options: MemberProfileControllerOptions['findFullProfile']): Promise<
        MemberProfileDto['MemberProfileController/FULL_MEMBER_PROFILE'] | undefined
    > {
        let _uri = '/members/';
        _uri += encodeURIComponent(options.memberId);
        _uri += '/profiles/full';
        return (await this.executor({uri: _uri, method: 'GET'})) as MemberProfileDto['MemberProfileController/FULL_MEMBER_PROFILE'] | undefined
    }
    
    async findProfile(options: MemberProfileControllerOptions['findProfile']): Promise<
        MemberProfileDto['MemberProfileController/DEFAULT_MEMBER_PROFILE'] | undefined
    > {
        let _uri = '/members/';
        _uri += encodeURIComponent(options.memberId);
        _uri += '/profiles/';
        return (await this.executor({uri: _uri, method: 'GET'})) as MemberProfileDto['MemberProfileController/DEFAULT_MEMBER_PROFILE'] | undefined
    }
    
    async save(options: MemberProfileControllerOptions['save']): Promise<
        Unit
    > {
        let _uri = '/members/profiles/';
        return (await this.executor({uri: _uri, method: 'PUT', body: options.body})) as Unit
    }
    
    async saveAvatar(options: MemberProfileControllerOptions['saveAvatar']): Promise<
        Unit
    > {
        let _uri = '/members/';
        _uri += encodeURIComponent(options.memberId);
        _uri += '/profiles/avatar/';
        return (await this.executor({uri: _uri, method: 'PUT'})) as Unit
    }
}

export type MemberProfileControllerOptions = {
    'findComplexProfiles': {
        readonly page?: number, 
        readonly size?: number, 
        readonly memberProfileInput: MemberProfileInput
    },
    'findFullProfile': {readonly memberId: string},
    'findProfile': {readonly memberId: string},
    'save': {readonly body: MemberProfileInput},
    'saveAvatar': {readonly memberId: string}
}

File: server/src/main/kotlin/cn/enaium/zensquare/controller/MemberProfileController.kt
/*
 * ZenSquare is an opensource forums
 *
 * Copyright (C) 2023  Enaium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package cn.enaium.zensquare.controller

import cn.dev33.satoken.annotation.SaIgnore
import cn.dev33.satoken.stp.StpUtil
import cn.enaium.zensquare.bll.error.ServiceException
import cn.enaium.zensquare.bll.service.ImageService
import cn.enaium.zensquare.model.entity.MemberProfile
import cn.enaium.zensquare.model.entity.by
import cn.enaium.zensquare.model.entity.input.MemberProfileInput
import cn.enaium.zensquare.repository.MemberProfileRepository
import cn.enaium.zensquare.util.checkId
import org.babyfish.jimmer.client.FetchBy
import org.babyfish.jimmer.kt.new
import org.babyfish.jimmer.sql.kt.fetcher.newFetcher
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile
import java.util.*

/**
 * member profile controller
 *
 * @author Enaium
 */
@RestController
class MemberProfileController(
    val memberProfileRepository: MemberProfileRepository,
    val imageService: ImageService
) {

    /**
     * Get member profile by member id
     *
     * @param memberId Member id
     * @return MemberProfile
     */
    @SaIgnore
    @GetMapping("/members/{memberId}/profiles/")
    fun findProfile(@PathVariable memberId: UUID): @FetchBy("DEFAULT_MEMBER_PROFILE") MemberProfile? {
        return memberProfileRepository.findByMemberId(memberId, DEFAULT_MEMBER_PROFILE)
    }

    /**
     * Get member profile by member id with full
     *
     * @param memberId Member id
     * @return MemberProfile
     */
    @SaIgnore
    @GetMapping("/members/{memberId}/profiles/full/")
    fun findFullProfile(@PathVariable memberId: UUID): @FetchBy("FULL_MEMBER_PROFILE") MemberProfile? {
        return memberProfileRepository.findByMemberId(memberId, FULL_MEMBER_PROFILE)
    }

    /**
     * Get member profiles
     *
     * @param page Page
     * @param size Size
     * @param memberProfileInput MemberProfileInput
     * @return Page<MemberProfile>
     */
    @GetMapping("/members/profiles/")
    fun findComplexProfiles(
        @RequestParam(defaultValue = "0") page: Int = 0,
        @RequestParam(defaultValue = "10") size: Int = 10,
        memberProfileInput: MemberProfileInput?
    ): Page<@FetchBy("DEFAULT_MEMBER_PROFILE") MemberProfile> {
        return memberProfileRepository.findAllByMemberProfile(PageRequest.of(page, size), memberProfileInput)
    }

    /**
     * Put member profile
     *
     * @param memberProfileInput MemberProfileInput
     */
    @PutMapping("/members/profiles/")
    fun save(@RequestBody memberProfileInput: MemberProfileInput) {
        //Check member profile owner
        memberProfileInput.memberId?.let {
            if (!(checkId(it) || StpUtil.hasRole("admin"))) {
                throw ServiceException(HttpStatus.FORBIDDEN)
            }
        }
        memberProfileRepository.save(memberProfileInput)
    }

    @PutMapping("/members/{memberId}/profiles/avatar/")
    @ResponseStatus(HttpStatus.OK)
    fun saveAvatar(@PathVariable memberId: UUID, file: MultipartFile) {
        //Check member profile owner
        if (!(checkId(memberId) || StpUtil.hasRole("admin"))) {
            throw ServiceException(HttpStatus.FORBIDDEN)
        }
        imageService.upload(file)
        memberProfileRepository.update(new(MemberProfile::class).by {
            this.memberId = memberId
            avatar = imageService.upload(file)
        })
    }

    companion object {
        /**
         * Default member profile fetcher
         */
        val DEFAULT_MEMBER_PROFILE = newFetcher(MemberProfile::class).by {
            nickname()
            avatar()
            createdTime()
            modifiedTime()
            member {
                createdTime()
            }
            role {
                name()
            }
        }

        /**
         * Full member profile fetcher
         */
        val FULL_MEMBER_PROFILE = newFetcher(MemberProfile::class).by {
            allScalarFields()
            member {
                allScalarFields()
                thread()
                reply()
                message()
                password(false)
            }
            role {
                allScalarFields()
            }
        }
    }
}